common/hashtable_key_context.h

```diff
@@ -69,16 +97,138 @@ namespace Carbon {
+class TranslatingKeyContext {
+ public:
+  // Derived types should provide one or more overloads that hide this function
+  // and perform translation for the key types which need it.
+  //
+  // For any key type, the context will check if there exists a callable
+  // `TranslateKey` function on the derived type. If so, that function will be
+  // called and the result used for hashing or comparison. If not, the key will
+  // be used directly. The derived type doesn't need to and shouldn't provide a
+  // default no-op overload. Instead, for any types that need no translation, it
+  // should ensure no overload is viable.
+  //
+  // Note that this function should be *hidden* by the derived overloads. It is
+  // provided here to help detect typos or misspellings or cases where no
+  // overload is provided at all.
+  template <typename TranslateKeyT>
+  auto TranslateKey(const TranslateKeyT& /*key*/) const -> int {
+    // A static_assert that will fail on any actual instantiation (it can't be
+    // instantiated with a void type). We have to make this dependent as
+    // Clang-16 will fail to compile even when the definition is never
+    // instantiated otherwise.
+    static_assert(
+        std::same_as<TranslateKeyT, void>,
+        "No `TranslateKey` overload was provided by the derived type!");
+  }
+
+  template <typename AnyKeyT>
+  auto HashKey(const AnyKeyT& key, uint64_t seed) const -> HashCode;
+
+  template <typename AnyKeyT, typename TableKeyT>
+  auto KeyEq(const AnyKeyT& lhs_key, const TableKeyT& rhs_key) const -> bool;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+//
+// Only implementation details below this point.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+namespace InternalHashtableEqDispatch {
+
+inline auto CarbonHashtableEq(const llvm::APInt& lhs, const llvm::APInt& rhs)
+    -> bool {
+  return lhs.getBitWidth() == rhs.getBitWidth() && lhs == rhs;
+}
+
+inline auto CarbonHashtableEq(const llvm::APFloat& lhs,
+                              const llvm::APFloat& rhs) -> bool {
+  return lhs.bitwiseIsEqual(rhs);
+}
+
+template <typename LeftT, typename RightT>
+inline auto CarbonHashtableEq(const LeftT& lhs, const RightT& rhs) -> bool
+  requires(requires {
+    { lhs == rhs } -> std::convertible_to<bool>;
+  })
+{
+  return lhs == rhs;
+}
+
+template <typename LeftT, typename RightT>
+inline auto DispatchImpl(const LeftT& lhs, const RightT& rhs) -> bool {
+  // This unqualified call will find both the overloads in our internal
+  // namespace above and ADL-found functions within an associated namespace for
+  // either `LeftT` or `RightT`.
+  return CarbonHashtableEq(lhs, rhs);
+}
+
+}  // namespace InternalHashtableEqDispatch
+
+template <typename LeftT, typename RightT>
+inline auto HashtableEq(const LeftT& lhs, const RightT& rhs) -> bool {
+  return InternalHashtableEqDispatch::DispatchImpl(lhs, rhs);
```

Comment:
You could remove a layer of indirection here, assuming this is the only call to `DispatchImpl`:
```suggestion
  using InternalHashtableEqDispatch::CarbonHashtableEq;
  CarbonHashtableEq(lhs, rhs);
```

Reply:
I wanted to keep the pattern as similar to what we use for `CarbonHashValue` as I could, so I replicated the dispatch routine... And we did end up using the dispatch there.

Still prefer skipping the indirection for now?

Reply:
I don't feel strongly about it. (One the one hand, the two-step approach is a bit more idiomatic because it's how people seem to be encouraged to call `std::swap`. On the other hand, your current approach can be made SFINAE-friendly whereas the `std::swap` approach can't.)

If this is more consistent with what you're doing elsewhere, let's leave it alone.